### PR TITLE
Add manpages for SSL_get_certificate, SSL_get_privatekey

### PR DESCRIPTION
**Pull request by hlandau**
This is as I understand these functions from reading the code.

### CODE REVIEW

File: doc/man3/SSL_get_certificate.pod [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ X509 *SSL_get_certificate(const SSL *s);
+ EVP_PKEY *SSL_get_privatekey(const SSL *s);
+
+=head1 DESCRIPTION
+
+SSL_get_certificate() returns a pointer to the B<X509> object representing the
+certificate which represents the local peer's identity. For servers, it returns
+the server's certificate. If a server has multiple certificates (for different
+algorithms, for example RSA and ECDSA), the certificate in actual use is
+returned. For clients, it returns the client certificate being used, if any.
```
Comment (mattcaswell) [inline doc/man3/SSL_get_certificate.pod:21]: I don't think this isn't quite correct. Before the handshake it returns the last certificate that was added to the SSL/SSL_CTX or NULL if none were added yet. During the handshake it selects a certificate to use for the connection and from that point on it returns the selected certificate.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by hlandau** [comment]
How's this?

File: doc/man3/SSL_get_certificate.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/man3/SSL_get_certificate.pod:21]: I still don't think this is quite right. I would merge the text you added under RETURN VALUES into this text. The text about "multiple certificates" at the moment is quite misleading in the case where it is called before the handshake. Perhaps adding something like "or the most recently added certificate if called before a certificate has been selected". Or something like that.

File: doc/man3/SSL_get_certificate.pod [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+SSL_get_privatekey() returns a pointer to the B<EVP_PKEY> object corresponding
+to the certificate returned by SSL_get_certificate().
+
+=head1 RETURN VALUES
+
+These functions return pointers to their respective objects, or NULL if no such
+object is available. Before the handshake has completed, the certificate
+returned is the one most recently added to the SSL object, or NULL if no
+certificate has yet been added. After the handshake has completed, the
+certificate returned is the one which was selected during the handshake.
```
Comment (mattcaswell) [inline doc/man3/SSL_get_certificate.pod:37]: I'd keep this more about about error conditions etc, e.g. "SSL_get_certifciate() returns a pointer to the certificate or a NULL if there is no certificate"

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by hlandau** [comment]
Updated again.

File: doc/man3/SSL_get_certificate.pod [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+certificate used as the local peer's identity.
+
+Multiple certificates can be configured; for example, a server might have both
+RSA and ECDSA certificates. The certificate which is returned by
+SSL_get_certificate() is determined as follows:
+
+=item
+
+If it is called before the handshake has completed, it returns the most recently
+added certificate, or NULL if no certificate has been added.
```
Comment (mattcaswell) [inline doc/man3/SSL_get_certificate.pod:27]: Strictly speaking the certificate selection happens *during* the handshake. This distinction is important in the case of callbacks that occur during the handshake (such as the tlsext_status_cb). So, the tlsext_statuc_cb will get the selected certificate (which is not necessarily the most recently added one) even though the handshake has not been completed yet.

**Comment by hlandau** [comment]
Updated.

File: doc/man3/SSL_get_certificate.pod [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+=head1 SYNOPSIS
+
+ #include <openssl/ssl.h>
+
+ X509 *SSL_get_certificate(const SSL *s);
+ EVP_PKEY *SSL_get_privatekey(const SSL *s);
+
+=head1 DESCRIPTION
+
+SSL_get_certificate() returns a pointer to a B<X509> object representing a
```
Comment (paulidale) [inline doc/man3/SSL_get_certificate.pod:17]: _an_ X509??

**Review by paulidale — APPROVED** [review]

File: doc/man3/SSL_get_certificate.pod (hunk 4 above)
Comment (hlandau) [reply doc/man3/SSL_get_certificate.pod]: Fixed. Thanks.

**Comment by hlandau** [comment]
Updated.

File: doc/man3/SSL_get_certificate.pod [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+example for the purpose of identifying a certificate's OCSP responder URL so
+that an OCSP response can be obtained.
+
+SSL_get_privatekey() returns a pointer to the B<EVP_PKEY> object corresponding
+to the certificate returned by SSL_get_certificate(), if any.
+
+=head1 RETURN VALUES
+
+These functions return pointers to their respective objects, or NULL if no such
+object is available.
```
Comment (mattcaswell) [inline doc/man3/SSL_get_certificate.pod:54]: It's probably worth mentioning that ownership of the returned object is retained by the SSL object (i.e. the ref count is not upped) - so users of this function should not free the returned object (or they should up the ref count themselves).


Aside: If we were to introduce such functions today we would call them `SSL_get0_*` to indicate this memory management behaviour

Comment (hlandau) [reply doc/man3/SSL_get_certificate.pod]: Thanks. Updated.

**Review by paulidale — APPROVED** [review]
Still approved (of course).

**Review by t8m — DISMISSED** [review]

**Comment by t8m** [comment]
The CI failure is relevant. Please fix.

File: doc/man3/SSL_get_certificate.pod [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+=head1 DESCRIPTION
+
+SSL_get_certificate() returns a pointer to an B<X509> object representing a
+certificate used as the local peer's identity.
+
+Multiple certificates can be configured; for example, a server might have both
+RSA and ECDSA certificates. The certificate which is returned by
+SSL_get_certificate() is determined as follows:
+
```
Comment (mattcaswell) [inline doc/man3/SSL_get_certificate.pod:23]: Needs an `=over 4` here

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good apart from the nit (which is causing the CI failure)

File: doc/man3/SSL_get_certificate.pod (hunk 6 above)
Comment (hlandau) [reply doc/man3/SSL_get_certificate.pod:23]: Fixed.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale still OK?

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to 3.0 and master branches. Thank you.

**Pull request closed without merging**